sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamMatchRel.java

```diff
@@ -0,0 +1,380 @@
+        this.regexPattern = regexPattern;
+      }
+
+      @ProcessElement
+      public void processElement(
+          @Element KV<Row, Iterable<Row>> keyRows, OutputReceiver<KV<Row, Iterable<Row>>> out) {
+        ArrayList<Row> rows = new ArrayList<>();
+        StringBuilder patternString = new StringBuilder();
+        for (Row i : keyRows.getValue()) {
+          rows.add(i);
+          // check pattern of row i
+          String patternOfRow = " "; // a row with no matched pattern is marked by a space
+          for (int j = 0; j < pattern.size(); ++j) {
+            CEPPattern tryPattern = pattern.get(j);
+            if (tryPattern.evalRow(i)) {
+              patternOfRow = tryPattern.toString();
+            }
+          }
+          patternString.append(patternOfRow);
+        }
+
+        Pattern p = Pattern.compile(regexPattern);
+        Matcher m = p.matcher(patternString.toString());
+        // if the pattern is (A B+ C),
+        // it should return a List three rows matching A B C respectively
+        if (m.matches()) {
+          out.output(KV.of(keyRows.getKey(), rows.subList(m.start(), m.end())));
+        }
+      }
+    }
+
+    private static class SortPerKey extends DoFn<KV<Row, Iterable<Row>>, KV<Row, Iterable<Row>>> {
+
+      private final Schema cSchema;
+      private final ArrayList<OrderKey> orderKeys;
+
+      public SortPerKey(Schema cSchema, RelCollation orderKeys) {
+        this.cSchema = cSchema;
+
+        List<RelFieldCollation> revOrderKeys = orderKeys.getFieldCollations();
+        Collections.reverse(revOrderKeys);
+        ArrayList<OrderKey> revOrderKeysList = new ArrayList<>();
+        for (RelFieldCollation i : revOrderKeys) {
+          int fIndex = i.getFieldIndex();
+          RelFieldCollation.Direction dir = i.getDirection();
+          if (dir == RelFieldCollation.Direction.ASCENDING) {
+            revOrderKeysList.add(new OrderKey(fIndex, false));
+          } else {
+            revOrderKeysList.add(new OrderKey(fIndex, true));
+          }
+        }
+
+        this.orderKeys = revOrderKeysList;
+      }
+
+      @ProcessElement
+      public void processElement(
+          @Element KV<Row, Iterable<Row>> keyRows, OutputReceiver<KV<Row, Iterable<Row>>> out) {
+        ArrayList<Row> rows = new ArrayList<Row>();
+        for (Row i : keyRows.getValue()) {
+          rows.add(i);
+        }
+        for (OrderKey i : orderKeys) {
+          int fIndex = i.getIndex();
+          boolean dir = i.getDir();
+          rows.sort(new SortComparator(fIndex, dir));
+        }
+        // TODO: Change the comparator to the row comparator:
+        // https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamSortRel.java#L373
+
+        out.output(KV.of(keyRows.getKey(), rows));
+      }
+
+      private class SortComparator implements Comparator<Row> {
+
+        private final int fIndex;
+        private final int inv;
+
```

Comment:
In my last PR (#12073), Rui suggested there is a row comparator available for use. I found it a private class in BeamSortRel. I just wonder if I could copy the code from there (maybe also and a reference?). What is the correct way of doing it?

Reply:
Hi you can update that class to public to use it.

Reply:
I just updated the implementation using the comparator in BeamSortRel. You will see it in my next commit!